File: 322. Coin Change/java/Solution322.java
import java.util.Arrays;

class Solution322 {
  public static void main(String[] args) {
    Solution sol = new Solution();
    // int[] coins = {1,2,5};
    // int amount = 11;
    int[] coins = {2,5,10,1};
    int amount = 27;
    System.out.println(sol.coinChange(coins, amount));
  }
}

class Solution {
  public int coinChange(int[] coins, int amount) {
    if (amount == 0) return 0;
    int[] dp = new int[amount + 1];
    Arrays.fill(dp, amount + 1);
    dp[0] = 0;
    for (int i = 1; i <= amount; i++) {
      for (int j = 0; j < coins.length; j++) {
        if (coins[j] <= i) {
          dp[i] = Math.min(dp[i], dp[i - coins[j]] + 1);
        }
      }
    }
    return dp[amount] > amount ? -1 : dp[amount];
  }
}

File: 322. Coin Change/go/solution322.go
package main

import "fmt"

func main() {
	coins := []int{1, 2, 5}
	amount := 11
	fmt.Println(coinChange(coins, amount))
}

const (
	MaxUint = ^uint(0)
	MinUint = 0
	MaxInt  = int(MaxUint >> 1)
	MinInt  = -MaxInt - 1
)

func coinChange(coins []int, amount int) int {
	cache := make(map[int]int)

	var dfs func(total int) int

	dfs = func(total int) int {
		if total == 0 {
			return 0
		}
		if total < 0 {
			return MaxInt
		}
		if val, ok := cache[total]; ok {
			return val
		}
		res := MaxInt
		for _, coin := range coins {
			if coin > total {
				continue
			}
			take := dfs(total - coin)
			if take != MaxInt {
				res = min(res, take+1)
			}
		}
		cache[total] = res
		return res
	}

	res := dfs(amount)
	if res == MaxInt {
		return -1
	}

	return res
}
